### PR TITLE
[maxtext] Add block-diffusion pre-training

### PR DESCRIPTION
**Pull request by ethannnnnn**
## Motivation

With block-diffusion primitives and block-causal attention available, MaxText
needs a focused pre-training integration before any completion fine-tuning,
rollout, distillation, replay, or RL behavior is introduced.

## Dependencies

This change follows and depends on:

- #4737 for model-independent corruption and target alignment; and
- #4743 for opt-in block-causal attention.

It has no Tunix dependency.

## Scope

This PR adds only:

- an opt-in `training_objective=block_diffusion` configuration;
- Hugging Face text pre-training corruption with separate clean targets,
  corruption masks, and target-loss masks;
- target-aligned loss handling for Linen and NNX models;
- explicit compatibility validation for the initial pre-training surface; and
- finite zero loss and gradients when an accumulated batch has no supervised
  positions; and
- a CI guard that skips secret-dependent notebook execution for fork and
  Dependabot pull requests, where GitHub withholds `HF_TOKEN`.

It does not add SFT, DPO, rollout, serving, distillation, replay, or RL.

## Training contract

The supported model contracts are:

- `same_position` logits with an `all_masked` canvas; and
- `shifted` logits with a `seed_and_mask` canvas.

Each nonempty eligible block keeps at least one supervised position. Clean
targets and segmentation metadata remain separate from the corrupted inputs.
Loss is normalized over `targets_loss_mask` positions only. Packing, SFT,
DPO, MTP, vocabulary tiling, audio, multimodal input, and non-Hugging-Face
pipelines fail validation for this initial scope.

## Compatibility

`training_objective=causal_lm` remains the default. Its next-token data
transform, segmentation-based loss mask, and autoregressive attention path are
unchanged.

The shared gradient-accumulation helper guards a zero total-weight denominator
only for the block-diffusion objective. The causal branch retains its original
operations, including the upstream Tunix accumulation-step divisor. Focused
regression tests verify both accumulation modes against the direct full-batch
loss and gradients.

## Tests

The audited current-main replay passed:

```text
270 passed, 58 skipped, 3 deselected, 92 subtests passed
```

The gate covers block-diffusion primitives, dense and TPU mask construction,
configuration, Hugging Face objective selection, padding and corruption,
Linen and NNX loss masks, causal-loss fallback, shaped batches, context
parallel metadata, and positive- and zero-weight gradient accumulation.

Three pre-existing Hugging Face iterator integration tests are deselected in
the local CPU gate because they require live `datasets.load_dataset` access.
The new objective-transform tests in that module pass independently.

All reachable changed production statements and branches are covered by this
gate: 122/122 executable lines and 70/70 branches. The one excluded clause is
the objective-specific packing error, which cannot execute because the
block-diffusion attention validation above it necessarily rejects packing
first.

With the CI JAX/libtpu toolchain, normalized optimized HLO for DeepSeek3,
Llama3 8B, and Qwen3 1.7B is byte-identical to the public PR parent. The
feature therefore requires no reference-HLO changes.

Pyink, Pylint, Python compilation, and `git diff --check` are also required by
the guarded publisher. Yamllint and an event-condition truth table cover the
fork-notebook guard; the public PR rerun must show the notebook job skipped and
the notebook aggregator passing.

The current-main replay retains upstream #4753's DeepSeek-V4 MoE changes and
passes the repository's unmodified Pylint policy, including `R1702`.

## Checklist

- [x] I performed a self-review of the focused pre-training boundary.
- [x] Existing autoregressive behavior remains the default and has regression coverage.
- [x] Unsupported training modes fail closed.
- [x] No post-training or RL implementation is included.

Design document:
https://docs.google.com/document/d/1N7KcCoAIErB2CV9EJ2G1u_mdN-AQgqwNUYSvM0mtqMI/edit

### CODE REVIEW

File: src/maxtext/input_pipeline/hf_data_processing.py [hunk 1]
```diff
@@ -42,6 +42,40 @@ def _get_pad_id(tokenizer):
...
+def _get_training_objective_transform(
+    config,
+    *,
+    shift,
+    use_dpo,
+    use_sft,
+    packing,
+    pad_id,
+    bos_token_id,
+):
```
Comment (hengtaoguo) [inline src/maxtext/input_pipeline/hf_data_processing.py]: Could you also add some doc string type hints to above args, for better understanding? Thanks

**Review by hengtaoguo — APPROVED** [review]
Thanks for adding the extensive test coverage. Please make sure to resolve the failing tests in CI before merging.

**Review by entrpn — APPROVED** [review]

**Review by shralex — APPROVED** [review]

**Comment by ethannnnnn** [comment]
@entrpn  It looks like the merge is blocked by 2 failing checks. The "pull ready" label was resolved manually but still shows as failing, and there's a google-internal check that failed whose link I don't have access to. Could you take a look?
> This site can’t be reached

**Comment by ethannnnnn** [comment]
> @entrpn It looks like the merge is blocked by 2 failing checks. The "pull ready" label was resolved manually but still shows as failing, and there's a google-internal check that failed whose link I don't have access to. Could you take a look?
> 
> > This site can’t be reached

followed with Juan offline and it looks like the out of date files cause the tests failure

**Review by hengtaoguo — APPROVED** [review]

File: src/maxtext/trainers/pre_train/train.py [hunk 2]
```diff
@@ -108,13 +109,35 @@ def loss_fn(model, config, data, dropout_rng, params, sparsity_state=None, is_tr
...
       data[k] = v[: config.micro_batch_size_to_train_on, :]
   else:
     for k, v in data.items():
       data[k] = v[: config.micro_batch_size_to_eval_on, :]
+  if is_block_diffusion:
+    targets_loss_mask = (data["targets_loss_mask"] != 0) & (data["targets_segmentation"] != 0)
+    target_positions = data.get("targets_position", data["inputs_position"])
+  else:
+    targets_loss_mask = None
+    target_positions = None
```
Comment (hengtaoguo) [inline src/maxtext/trainers/pre_train/train.py:140]: Added two-lines to resolve [internal error](https://screenshot-v2.corp.google.com/1t0rddiq16ep8)

```
INFO Found 5 type errors
ERROR `target_positions` may be uninitialized [unbound-name]
   --> [third_party/py/maxtext/src/maxtext/trainers/pre_train/train.py:194](https://cs.corp.google.com/piper///depot/google3/third_party/py/maxtext/src/maxtext/trainers/pre_train/train.py?l=194&ws=tap-presubmit-server/676152626&snapshot=2):13
    |
194 |             target_positions,
    |             ^^^^^^^^^^^^^^^^
    |
ERROR `targets_loss_mask` may be uninitialized [unbound-name]
   --> [third_party/py/maxtext/src/maxtext/trainers/pre_train/train.py:216](https://cs.corp.google.com/piper///depot/google3/third_party/py/maxtext/src/maxtext/trainers/pre_train/train.py?l=216&ws=tap-presubmit-server/676152626&snapshot=2):23
    |
216 |         xent = xent * targets_loss_mask
    |                       ^^^^^^^^^^^^^^^^^
    |
```

**Comment by hengtaoguo** [comment]
After some (1) internal BUILD updates, (2) rebase & resolve conflicts, and (3) small patches for type enhancement, we are seeing internal tests [all green](https://screenshot-v2.corp.google.com/7m5cgefaqvo5o) now. This is good news!

But now an unrelated error in `test_tpu_flash_attention_usp_hlo_uses_all_to_all_and_permute ` is [failing](https://github.com/AI-Hypercomputer/maxtext/actions/runs/31770198650/job/94674732886) at maxtext head. We probably need code owner to solve this. I'm also considering merging this PR since the failures are pre-exsiting and unrelated. We will keep an eye on it, thanks again for the contribution.
